### ISSUE TITLE
Should we / can we make all "getelementptr inbounds" into "getelementptr nowrap"?

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#350]
The "inbounds" semantics of `offset` are [notoriously tricky and confusing](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-lang.2Fwg-unsafe-code-guidelines/topic/inbounds.20offsets.20can.20leave.20the.20provenance.20region). From what I hear from @nikic, the "inbounds" part of them is also [not nearly as useful as one might think](https://discourse.llvm.org/t/question-about-getelementptr-inbounds-with-offset-0/62533/9), and the main payoff is being sure that the pointer is not wrapped around either end of the address space.

So... is there a chance that we could significantly simplify the language at acceptable cost for analyses by changing the rules of `offset` (and all other "inbounds" offsets that the language does implicitly, like when applying place projections) such that the only case of UB here is overflow wrapping around the address space (both below `0` and above `usize::MAX`)? I think that would be great, but of course we have to be careful not to give up too much information here. (That said, we *do* have a ton of information of the form "this pointer is dereferenceable for size N", which conveys bounds information much more directly than `getelementptr inbounds`.)

However, we'd probably need LLVM support for this, adding some sort of `getelementptr nowrap`. (There *is* the possible alternative of using plain `getelementptr`, and upgrading that to `inbounds` whenever we can derive from other information that the pointer is indeed dereferenceable for a sufficiently large memory range. I am not sure how tricky that would be to implement though.)

So I wonder, @nikic, do you think that would be a reasonable and realistic option? And everyone, do you think that would be a reasonable semantics to shoot for?

**Comment by Lokathor** [comment]
Just saying "you can't wrap the address space" is extremely teachable.

**Comment by scottmcm** [comment]
I always thought this was as much about aliasing information as anything.  Is `GEPi` not actually important for that because we get all the information we need about it from provenance anyway?

**Comment by Lokathor** [comment]
Yeah two pointers that each point to a separate stack object still can't index oob and access each other because they'd have separate provenance.

**Comment by eddyb** [comment]
If GEP was replaced by an untyped "pointer offset with integer dot product" operation (i.e. a sequence of constant strides and runtime indices, to cover the "nested arrays" case, if they do want to keep it as one instruction), it would be great to reuse `nuw`/`nsw` flags from arithmetic operations.

(NUW/NSW standing for "No Unsigned/Signed Wrap", i.e. opting into to unsigned/signed overflow being UB)

Alternatively, if LLVM had `add nuw`/`sub nuw` between `ptr` and an integer, that could allow for `add`/`sub` methods on Rust pointers that aren't limited by `isize` like `offset` is (though I'm not sure we want to even dream about such things, given how much legacy there is around the whole "`ptrdiff_t` is one bit too small" issue).

**Comment by RalfJung** [comment]
Even on the LLVM side there seems to be [some desire for a `getelementptr nowrap`](https://discourse.llvm.org/t/retain-fact-that-index-offsets-for-gep-are-non-negative-for-size-t-uint64-t-offsets/65974/16?u=ralfjung), though it seems like that proposal has not moved in a while.

**Comment by RalfJung** [comment]
I learned in the mean time that `getelementptr inbounds` in LLVM doesn't actually require the allocation to be *live*. In my interpretation of Rust's rules (and in Miri's implementation), we do require liveness. So we could weaken our `offset` rules a bit by dropping the liveness requirement. But honestly this doesn't solve the main pain points here so I don't think it's worth it.

**Comment by RalfJung** [comment]
We had a t-opsem meeting on this subject; see https://github.com/rust-lang/opsem-team/issues/10 for a summary.

**Comment by GoldsteinE** [comment]
There’s now a `&raw const (*ptr).field` pattern for projections, which is `inbounds`, but it’s actually rather hard to find documentation stating that. (The closest available thing is [documentation for `addr_of!()`](https://doc.rust-lang.org/stable/std/ptr/macro.addr_of.html), but you first need to know that `&raw const` is the same thing).

**Comment by RalfJung** [comment]
It's documented at https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html:

> Performing a place projection that violates the requirements of [in-bounds pointer arithmetic](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.offset).

**Comment by saethlin** [comment]
@GoldsteinE Did you find documentation for `&raw const (*ptr).field` that didn't mention this requirement? I'm asking, because to be honest I'm not sure what documentation for that expression exists.

**Comment by GoldsteinE** [comment]
@saethlin There’s some documentation on a raw pointer type page: https://doc.rust-lang.org/stable/std/primitive.pointer.html#3-create-it-using-raw
